Add support for Log4j 2's ThreadContext

Marcono1234 opened the pull request:
Resolves #188

The code orientates itself to the existing SLF4J MDC one.

Noteworthy:
- Renames the SLF4J MDC module (but not its `artifactId`) to make the distinction between Log4j 1 & 2 (which use the term MDC as well) clearer
- Log4j 2 ThreadContext module currently supports Java 5 even though latest Log4j 2 versions require Java 8; this might allow usage with older Log4j 2 versions, but I have not tested it (relates to #112)
- Log4j 2 `ThreadContextManager.clear()` clears the ThreadContext of the current thread (see also discussion in #204)
- Log4j 2 `ThreadContextManager.getActiveContext().close()` does nothing (similar to SLF4J MDC one)
- Propagating Log4j 2 ThreadContext to a different thread will add the data on top of the existing one instead of replacing it (see also discussion in #202)

sjoerdtalsma commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+     *
+     * @param value non-{@code null} data for the {@code ThreadContext}
+     * @return The new <em>active</em> context containing the specified value
+     * which should be closed by the caller at the end of its lifecycle.
```
Maybe add _from the same thread_

sjoerdtalsma commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+ * Represents a snapshot of the data of the Log4j 2 {@link ThreadContext} of a
+ * specific thread at a certain point in the past.
+ */
+public class ThreadContextData {
```
When multiple context managers are applied it may be useful to know this is Log4j2's data?
Perhaps create an interface `Log4j2ThreadContext` with implementation `Log4j2ThreadcontextData` ?

Marcono1234 replied:
I am not sure if having an interface is worth it, see https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577189357. We are not expecting the user to provide their own implementation of that class, are we?

sjoerdtalsma commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+ *
+ * @see <a href="https://logging.apache.org/log4j/2.x/manual/thread-context.html">Log4j 2 Thread Context manual</a>
+ */
+public class ThreadContextManager implements ClearableContextManager<ThreadContextData> {
```
I prefer `Log4j2ThreadContextManager` in case multiple context managers are used.

Marcono1234 replied:
Prefixed classes with `Log4j2` and changed package name from `log4j2` to `log4j2.threadcontext` (similar to how you renamed the SLF4J MDC package).

sjoerdtalsma commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/package-info.java:
```diff
@@ -0,0 +1,21 @@
+ * Propagate the {@linkplain org.apache.logging.log4j.ThreadContext Log4j 2 ThreadContext}
+ * data from one thread to another.
+ */
+package nl.talsmasoftware.context.log4j2;
```
Missing newline at end of file?

sjoerdtalsma commented on log4j2-propagation/src/main/resources/META-INF/services/nl.talsmasoftware.context.ContextManager:
```diff
@@ -0,0 +1 @@
+nl.talsmasoftware.context.log4j2.ThreadContextManager
```
Missing newline at end of file?

sjoerdtalsma commented on slf4j-mdc-propagation/pom.xml:
```diff
@@ -27,7 +27,7 @@
     <!-- Artifact identification -->
     <artifactId>mdc-propagation</artifactId>
-    <name>Propagation of MDC</name>
+    <name>Propagation of SLF4J MDC</name>
```
Please also rename artifactId to `slf4j-mdc-propagation` to maintain consistency.

We can add a `mdc-propagation` module to the relocations that redirects users to the new artifact.

Marcono1234 replied:
Should we rename it then to `slf4j-propagation` in case this library covers aspects other than MDC in the future as well (this was also the reason why I chose `log4j2-propagation`)?

sjoerdtalsma replied:
I'd prefer it if the rename were tackled in a separate PR, so there are less different topics to discuss in this one. (I created #209 for the Slf4j module rename)

sjoerdtalsma reviewed (changes requested):
Thanks for your great work!
I did an initial review from the online github interface but will check out your branch to have a closer look from the IDE.

Please don't be overwhelmed by the number of comments; I tried to be constructive in my criticism and I like the general quality of this PR, which _again_ is greatly appreciated!

Marcono1234 commented on log4j2-propagation/README.md:
```diff
@@ -0,0 +1,41 @@
+Adding the `log4j2-propagation` jar to your classpath
+is all that is needed to let the [Thread Context][thread-context] 
+from the [Log4j 2 library][log4j2] be automatically included into
+the `ContextSnapshot`.
```
How exactly do you mean that?
Log4j 2 on its own is completely unrelated to SLF4J, it does not implement the SLF4J interface. So someone using only Log4j 2 might not be interested in information about SLF4J.

There is however a [Log4j to SLF4J Adapter](https://logging.apache.org/log4j/2.x/log4j-to-slf4j/index.html) (as well as [Log4j 2 SLF4J Binding](https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/), which is SLF4J to Log4j 2). Do you mean for these 'bridges' a comment would be useful (though I don't actually know how well they work with MDC and ThreadContext)?

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+     * @param data data to apply, may be {@code null}
+     * @param overwrite whether all existing data should overwritten
+     */
+    static void applyToCurrentThread(ThreadContextData data, boolean overwrite) {
```
These methods are only intended for usage by `ThreadContextManager`. It might also be possible to have this logic only in `ThreadContextManager`, work with the public methods of `ThreadContextData` and have its constructor be package-private, but I felt the logic belongs closer to `ThreadContextData` than to `ThreadContextManager`.

As for why these methods are package-private: Most of them provide internal functionality already provided (indirectly) through `ThreadContextManager` or through Log4j 2's ThreadContext. Providing it publicly might be confusing and lead to incorrect usage.
Making the whole class package-private is not an option because it is publicly exposed through `ThreadContextManager` and would therefore cause usability issues when someone tries to use `ThreadContextManager`.

I don't think making it an interface fits really well because it is not expected that the user implements or subclasses this class. Because the package-private methods are not accessible to the user and are not shown in the javadoc either, I don't think that this is an issue.

Would it be alright for you to leave the class as is? If you want I can add a comment (not part of the documentation) to the class describing the rationale I described here.

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+     * @param data data to apply, may be {@code null}
+     * @param overwrite whether all existing data should overwritten
+     */
+    static void applyToCurrentThread(ThreadContextData data, boolean overwrite) {
```
There is an instance method `applyToCurrentThread()` as well, which could be `private`, but is package-private for testing (maybe I should rewrite the tests to not use that method though).
The reason why this method exists is similar to what I described in https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577189357, I wanted to have logic related to manipulation of Log4j 2's `ThreadContext` only in `ThreadContextData` because I feel it belongs more there than to `ThreadContextManager`. And because the `data` can be `null`, only having an instance method would not work.

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+     * Applies the data on top of the {@code ThreadContext} of the current thread.
+     * Conflicting entries are overwritten but all other existing data is kept.
+     */
+    void applyToCurrentThread() {
```
See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577191621

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    private ThreadContextData(Map<String, String> contextMap, List<String> contextStack) {
+        this.contextMap = Collections.unmodifiableMap(contextMap);
```
The methods `getContextMap()` and `getContextStack()` publicly expose the values of these fields and the user should not be able to manipulate them; not even the `ThreadContextData` class itself should be able to manipulate them, since the data represent an unmodifiable snapshot.

It sould also be possible to wrap them in unmodifiable collections every time `getContextMap()` and `getContextStack()` is called, but that would only create unnecessary overhead.

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    static ThreadContextData fromCurrentThreadContext() {
+        // Get a copy of context map and context stack
+        return new ThreadContextData(ThreadContext.getContext(), ThreadContext.getImmutableStack().asList());
```
I find it personally easier to work with a `List` than with `ContextStack`.
Additionally opposed to what the name suggests the result of `getImmutableStack()` can actually be manipulated: `ThreadContext.getImmutableStack().asList().add("test")`
I assume that might be a bug, but I have not reported it to Log4j 2 yet. To be on the safe side I rather wanted to rely on the JDK methods for making a list unmodifiable.

Originally I did not expose these fields at all, but then I thought that would make the `ThreadContextData` class look rather useless from outside.
However, in the end I think the user should use Log4j 2's `ThreadContext` directly and not the `ThreadContextData` for accessing the data.

Should I change it anways?

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * Singleton instance of this class.
+     */
+    public static final ThreadContextManager INSTANCE = new ThreadContextManager();
```
`provider()` is tightly coupled with `ServiceLoader` that I was not sure whether it is good to point the user to that as well. Especially since you normally expect a method like `getInstance()` to return a singleton instance, and not `provider()`.

Do you want this do be changed anyways?

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+ * As with all manager implementations of this library there is usually no need to directly
+ * interact with the manager classes. Instead Java's {@link ServiceLoader} makes sure they
+ * are loaded as services. If an instance of this class is needed nonetheless it can be obtained
+ * through the field {@link #INSTANCE}.
```
See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577200546

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+     *
+     * @deprecated
+     *      This constructor only exists for usage by {@link ServiceLoader}. The singleton instance
+     *      obtained from {@link #INSTANCE} should be used instead.
```
See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577200546

sjoerdtalsma commented:
Just to touch base, I have been rather busy lately with many things on my mind.
But let me ensure you, this PR has not been forgotten by me and it's on my list of things to do.

sjoerdtalsma commented:
Could you please split this PR into the Log4j2 module only and (optionally) create separate PR's for the other concerns?
It is a bit much for me to review all in a single PR and I would like to be able to focus on the Log4j2 implementation, so we can have a clean first version.

sjoerdtalsma commented:
I'll create a PR to rename the MDC module (with a new relocation in place for it)

Marcono1234 commented:
> Could you please split this PR into the Log4j2 module only and (optionally) create separate PR's for the other concerns?

Yes, I can do that. Though if you don't mind, it would like to wait for your pull request to be merged first and then rebase this pull request onto `develop`.

Marcono1234 commented on context-propagation-java5/src/main/java/nl/talsmasoftware/context/ContextSnapshot.java:
```diff
@@ -49,7 +49,7 @@
      * is a safe way to propagate context snapshots without having to worry about closing them.
      *
-     * @return A new reactivationcontext with the snapshot values that will be valid until closed
+     * @return A new reactivation-context with the snapshot values that will be valid until closed
```
Will be resolved in #217.

Marcono1234 commented on context-propagation-java5/src/main/java/nl/talsmasoftware/context/clearable/Clearable.java:
```diff
@@ -15,8 +15,10 @@
+
 /**
- * Interface that marks an to be 'clearable'.
+ * Interface that marks a {@link ContextManager} to be 'clearable'.
```
Will be resolved in #217.

Marcono1234 commented on log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:
```diff
@@ -0,0 +1,167 @@
+    public static final ThreadContextManager INSTANCE = new ThreadContextManager();
+
+    /**
+     * Returns the singleton instance.
```
I think it does not hurt to mention this implementation detail; especially since the reason why this method exists is that there is a singleton instance. Otherwise the no-arg constructor would suffice for `ServiceLoader`.

Marcono1234 commented:
Sorry for the delay; I have rebased the changes onto #217 (so that should be merged first), and addressed some of your review comments (which I then marked as resolved).

sjoerdtalsma commented:
Do you want to rebase on _develop_ again? I merged #217 just now.
I think it would be a good idea to focus on the new `log4j2` module and leave all else out of this PR.

sjoerdtalsma commented:
I have some issues still, but I can work out those myself after I've merged this.
Let me thank you once more sincerely for your work and your patience with code review 😉.

Marcono1234 commented:
Thanks for merging the changes!

sjoerdtalsma commented:
> Thanks for merging the changes!

Thanks for contributing them!
